edge-hub/watchdog/src/main.rs

```diff
@@ -0,0 +1,85 @@
+mod child;
+
+use std::{
+    io::Error,
+    sync::atomic::{AtomicBool, Ordering},
+    sync::Arc,
+    thread,
+};
+
+use anyhow::{Context, Result};
+use child::run;
+use signal_hook::{iterator::Signals, SIGINT, SIGTERM};
+use tracing::{error, info, subscriber, Level};
+use tracing_subscriber::fmt::Subscriber;
+
+fn main() -> Result<()> {
+    init_logging();
+    info!("Starting watchdog");
+
+    let should_shutdown = register_shutdown_listener()
+        .context("Failed to register sigterm listener. Shutting down.")?;
+
+    let broker_handle = run(
+        "MQTT Broker".to_string(),
+        "/usr/local/bin/mqttd".to_string(),
+        "-c /tmp/mqtt/config/production.json".to_string(),
+        Arc::clone(&should_shutdown),
+    )?;
+
+    let edgehub_handle = match run(
+        "Edge Hub".to_string(),
+        "dotnet".to_string(),
+        "/app/Microsoft.Azure.Devices.Edge.Hub.Service.dll".to_string(),
+        Arc::clone(&should_shutdown),
+    ) {
+        Ok(handle) => Some(handle),
+        Err(e) => {
+            should_shutdown.store(true, Ordering::Relaxed);
+            error!("Could not start edgehub process. {}", e);
+            None
+        }
+    };
+
+    if let Err(e) = broker_handle.join() {
+        should_shutdown.store(true, Ordering::Relaxed);
+        error!("Failure while running broker process. {:?}", e)
+    }
+    info!("Successfully stopped broker process");
+
+    edgehub_handle.map(|handle| {
```

Comment:
```suggestion
    edgehub_handle.map(|handle| handle.join()).with_context(|| {
        should_shutdown.store(true, Ordering::Relaxed);
        format!("Failure while running edgehub process. {:?}", e)
    }?;
    info!("Successfully stopped edgehub process");
```

Reply:
This wouldn't compile because handle.join() returns (), for which std::ops::try is not implemented.

```
the `?` operator can only be applied to values that implement `std::ops::Try`
the trait `std::ops::Try` is not implemented for `()`
required by `std::ops::Try::into_result`
```

I am not sure why though because handle.join() should return Result<()> which should be fine?